fableous-fe/src/components/Canvas.tsx

```diff
@@ -18,417 +18,609 @@ import { ControllerRole, ToolMode, WSMessage, WSMessageType } from "../Data";
+          ...prev,
+          [id]: {
+            text,
+            fontSize,
+            x1,
+            y1,
+            x2,
+            y2,
+          } as TextShape,
+        }));
+        if (role !== ControllerRole.Hub) {
+          const [normX, normY] = scaleDownXY(x, y);
+          const [normFontSize] = scaleDownXY(fontSize, 0);
+          wsRef.current?.send(
+            JSON.stringify({
+              role,
+              type: WSMessageType.Text,
+              data: {
+                x1: normX,
+                y1: normY,
+                id,
+                text,
+                width: normFontSize,
+              },
+            } as WSMessage)
+          );
+        }
+      },
+      [canvasRef, getTextBounds, role, scaleDownXY, wsRef]
+    );
+
+    const refreshText = useCallback(() => {
       const ctx = canvasRef.current.getContext(
         "2d"
       ) as CanvasRenderingContext2D;
-      ctx.font = `${fontSize * SCALE}px Arial`;
-      ctx.textAlign = "center";
-      ctx.fillText(message, x, y);
-      if (role !== ControllerRole.Hub) {
-        const [normX, normY] = scaleDownXY(x, y);
-        const [normFontSize] = scaleDownXY(fontSize, 0);
-        wsRef.current?.send(
-          JSON.stringify({
-            role,
-            type: WSMessageType.Text,
-            data: {
-              x1: normX,
-              y1: normY,
-              width: normFontSize,
-              text: message,
-            },
-          } as WSMessage)
+      if (!ctx) return;
+      const { width, height } = canvasRef.current;
+      ctx.clearRect(0, 0, width, height);
+      Object.entries(textShapes).forEach(([id, shape]) => {
+        ctx.font = `${shape.fontSize * SCALE}px Arial`;
+        ctx.fillText(
+          shape.text,
+          (shape.x1 + shape.x2) / 2,
+          (shape.y1 + shape.y2) / 2
+        );
+        if (parseInt(id, 10) === editingText) {
+          ctx.beginPath();
+          ctx.strokeStyle = "#00aaaa";
+          ctx.rect(
+            shape.x1 - SELECT_PADDING,
+            shape.y1 - SELECT_PADDING,
+            shape.x2 - shape.x1 + 2 * SELECT_PADDING,
+            shape.y2 - shape.y1 + 2 * SELECT_PADDING
+          );
+          ctx.stroke();
+          ctx.closePath();
+        }
+      });
+    }, [canvasRef, editingText, textShapes]);
+
+    const interactCanvas = (x: number, y: number, editing: boolean) => {
+      let clickedId: number | undefined;
+      let clicked: TextShape | undefined;
+      Object.entries(textShapes).forEach(([id, shape]) => {
+        if (
+          !clicked &&
+          x >= shape.x1 - SELECT_PADDING &&
+          x <= shape.x2 + SELECT_PADDING &&
+          y >= shape.y1 - SELECT_PADDING &&
+          y <= shape.y2 + SELECT_PADDING
+        ) {
+          clickedId = parseInt(id, 10);
+          clicked = shape;
+        }
+      });
+      if (editing) {
+        if (clickedId && clicked) {
+          setEditingText(clickedId);
+          console.log(clicked);
```

Comment:
do u want to remove occurences of console.logs in this file? except ones that print error in a catch block

Reply:
done